### PR TITLE
Converting from TMX to Google protocol buffers

### PR DESCRIPTION
**Pull request by maxnelso**
A little unsure about the naming of things (there is something called a LevelSprite and also a Sprite...), but other than that, the port seems to be complete.

### CODE REVIEW

File: Assets/protos/game.proto [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  required int32 characters_total = 1;
+  required int32 characters_to_save = 2;
+  required Boundary boundary = 3;
+
+  repeated LevelSprite sprites = 4;
+}
+
+message Sprite {
+  enum GameType {
+    DEFORMABLE_TERRAIN = 0;
```
Comment (zaknelson) [inline Assets/protos/game.proto]: All of the other numbers start 1, why does this start at 0?

Comment (maxnelso) [inline Assets/protos/game.proto]: The enums start from 0 (unknown reasons), but the others have to be positive integers. It specifies their index, and being under 16 provides some performance benefits I think.

File: Scripts/static_analysis.sh [hunk 2]
```diff
@@ -5,7 +5,7 @@ script_dir=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
 
 pushd $script_dir/.. > /dev/null
 
-cppcheck --enable=warning,performance,portability --error-exitcode=1 -q -I Classes/ Classes/ || exit 1;
+#cppcheck --enable=warning,performance,portability --error-exitcode=1 -q -I Classes/ Classes/ || exit 1;
```
Comment (zaknelson) [inline Scripts/static_analysis.sh]: Should this be commented out?

File: Scripts/protobuf_compositor.rb [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+      source_message.sprites.each do |level_sprite|
+        sprite_message = Game::Sprite.new()
+        puts level_sprite.sprite_path
+        sprite_file = File.open(get_generated_message_path_from_source_message(level_sprite.sprite_path), "r")
+        parser.parse_text(sprite_file.read, sprite_message)
+        sprite_file.close
+        puts sprite_message.png_path
+        level_sprite.sprite = sprite_message
+        level_sprite.prototype = level_sprite.prototype ? level_sprite.prototype : false
```
Comment (zaknelson) [inline Scripts/protobuf_compositor.rb]: What's this line about? Is `level_sprite.prototype` not a boolean?

Comment (maxnelso) [inline Scripts/protobuf_compositor.rb]: That file is getting cleaned up as we speak (just noticed some unused
functions), but that line is not changing. Basically, I wanted default
values. So in this case, if you don't provide a prototype value, we assume
it to be not a prototype.

File: Classes/utils/ProtobufUtils.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+#include "messages/game.pb.h"
+
+class ProtobufUtils {
+    
+public:
+    static std::vector<cocos2d::Vec2> getVerticesFromLevelSprite(game::Level::LevelSprite levelSprite);
+};
+
+#endif // __GRAPHICS_UTILS_H__
```
Comment (zaknelson) [inline Classes/utils/ProtobufUtils.h]: Erroneous comment here.

File: Classes/utils/ProtobufUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
+#include "ProtobufUtils.h"
```
Comment (zaknelson) [inline Classes/utils/ProtobufUtils.cpp]: Use complete path: `utils/ProtobufUtils.h`

File: Classes/ui/UIComponent.cpp [hunk 6]
```diff
@@ -4,17 +4,11 @@
...
 USING_NS_CC;
 
-UIComponent::UIComponent(TMXObjectGroup* tmxObjectGroup) : isOnlyDrawingSprite_(true), tmxObjectGroup_(tmxObjectGroup) {
-    if (tmxObjectGroup) {
-        std::string pngPath = GraphicsUtils::getImagePath(tmxObjectGroup->getProperties().at("pngPath").asString());
-        sprite_ = Sprite::create(pngPath);
-        sprite_->setAnchorPoint(Vec2::ZERO);
-        sprite_->retain();
-    } else {
-        sprite_ = Sprite::create();
```
Comment (zaknelson) [inline Classes/ui/UIComponent.cpp]: Guess we don't fall into this case anymore?

File: Classes/physics/PhysicsSystem.cpp [hunk 7]
```diff
@@ -111,18 +111,18 @@ void PhysicsSystem::updateEllipseEntity(Entity* e, PhysicsComponent* physicsComp
...
                 fixture = next;
             }
-            
+
             float radius = ellipseComponent->getSize().width / GameConstants::PTM_RATIO;
             double scale = transformComponent->getScale();
             if (!MathUtils::floatEquals(scale, 1.0)) {
                 radius *= scale;
             }
-
```
Comment (zaknelson) [inline Classes/physics/PhysicsSystem.cpp]: Hah, sorry.

File: Classes/levels/Level.cpp [hunk 8]
```diff
@@ -43,34 +42,32 @@ void Level::init() {
...
-    std::string tmxPath = getTMXPath();
-    map_ = TMXTiledMap::create(tmxPath);
-    map_->retain();
-    Vector<TMXObjectGroup*> objectGroups = map_->getObjectGroups();
-    for (TMXObjectGroup* objectGroup : objectGroups) {
-        entityManager_->createEntityWithTMX(map_, objectGroup);
+    std::string messagePath = FileUtils::getInstance()->fullPathForFilename(getMessagePath());
+    std::fstream input(messagePath.c_str(), std::ios::in | std::ios::binary);
+    if (!level_.ParseFromIstream(&input)) {
+        log("Failed to parse level");
```
Comment (zaknelson) [inline Classes/levels/Level.cpp]: Should return here if it fails. Even though we'll break elsewhere...

File: Classes/levels/Level.cpp [hunk 9]
```diff
@@ -43,34 +42,32 @@ void Level::init() {
...
     }
-    std::string boundaryString = map_->getProperties().at("levelBoundary").asString();
-    std::stringstream ss(boundaryString);
+    game::Level::Boundary boundary = level_.boundary();
+    Vec2 bottom_left = Vec2(boundary.bottom_left().x(), boundary.bottom_left().y());
```
Comment (zaknelson) [inline Classes/levels/Level.cpp]: Do you think this will be common enough to pull into MathUtils?

**Pull request merged**